File: src/main/java/de/schramm/royalbash/tdd/controller/service/core/Spot.java
package de.schramm.royalbash.tdd.controller.service.core;

import lombok.Builder;
import lombok.Value;
import lombok.val;

import java.util.Optional;

@Value
@Builder(toBuilder = true)
public class Spot {

    private final Creature creature;

    public Optional<Creature> getCreature() {
        return Optional.ofNullable(creature);
    }

    public Spot place(Creature creature) {

        if(!creature.canBePlacedOnSpot()) {
            return this;
        }

        return this.toBuilder()
                .creature(creature)
                .build();
    }

    Spot updateCreature(Creature oldCreature, Creature newCreature) {

        val updatedCreature = getCreature()
                .map(creature -> creature.equals(oldCreature) ? newCreature : creature)
                .orElse(null);

        return this.toBuilder()
                .creature(updatedCreature)
                .build();
    }

    Spot removeCreature(Creature creature) {
        return this.toBuilder()
                .creature(creature.equals(this.creature) ? null : this.creature)
                .build();
    }
}

File: src/main/kotlin/de/schramm/royalbash/domain/Spot.kt
package de.schramm.royalbash.domain

import java.util.*

data class Spot(
        val id: String,
        private val creature: Creature? = null) {

    fun getCreature(): Optional<Creature> {
        return Optional.ofNullable(creature)
    }

    fun place(creature: Creature): Spot {
        return if (!creature.placeableOnSpot) {
            this
        } else copy(creature = creature)
    }

    internal fun updateCreature(oldCreature: Creature, newCreature: Creature): Spot {

        val updatedCreature = getCreature()
                .map { creature -> if (creature == oldCreature) newCreature else creature }
                .orElse(null)

        return copy(creature = updatedCreature)
    }

    internal fun removeCreature(creature: Creature): Spot {
        return copy(creature = if (creature == this.creature) null else this.creature)
    }
}
